pkg/kubelet/volume/cache/actual_state_of_world.go

```diff
@@ -0,0 +1,586 @@
+	// cache indicating the volume has been successfully unmounted from the pod.
+	// If a pod with the same unique name does not exist under the specified
+	// volume, this is a no-op.
+	// If a volume with the name volumeName does not exist in the list of
+	// attached volumes, this is a no-op.
+	DeletePodFromVolume(podName volumetypes.UniquePodName, volumeName api.UniqueVolumeName)
+
+	// DeleteVolume removes the given volume from the list of attached volumes
+	// in the cache indicating the volume has been successfully detached from
+	// this node.
+	// If a volume with the name volumeName does not exist in the list of
+	// attached volumes, this is a no-op.
+	// If a volume with the name volumeName exists and its list of mountedPods
+	// is not empty, an error is returned.
+	DeleteVolume(volumeName api.UniqueVolumeName) error
+
+	// PodExistsInVolume returns true if the given pod exists in the list of
+	// mountedPods for the given volume in the cache, indicating that the volume
+	// is attached to this node and the pod has successfully mounted it.
+	// If a pod with the same unique name does not exist under the specified
+	// volume, false is returned.
+	// If a volume with the name volumeName does not exist in the list of
+	// attached volumes, a volumeNotAttachedError is returned indicating the
+	// given volume is not yet attached.
+	// If a the given volumeName/podName combo exists but the value of
+	// remountRequired is true, a remountRequiredError is returned indicating
+	// the given volume has been successfully mounted to this pod but should be
+	// remounted to reflect changes in the referencing pod. Volumes like
+	// Downward API, depend on this to update the contents of the volume.
+	// All volume mounting calls should be idempotent so a second mount call for
+	// volumes that do not need to update contents should not fail.
+	PodExistsInVolume(podName volumetypes.UniquePodName, volumeName api.UniqueVolumeName) (bool, error)
+
+	// GetMountedVolumes generates and returns a list of volumes and the pods
+	// they are successfully attached and mounted for based on the current
+	// actual state of the world.
+	GetMountedVolumes() []MountedVolume
+
+	// GetMountedVolumesForPod generates and returns a list of volumes that are
+	// successfully attached and mounted for the specified pod based on the
+	// current actual state of the world.
+	GetMountedVolumesForPod(podName volumetypes.UniquePodName) []MountedVolume
+
+	// GetAttachedVolumes generates and returns a list of all attached volumes.
+	GetAttachedVolumes() []AttachedVolume
+
+	// GetUnmountedVolumes generates and returns a list of attached volumes that
+	// have no mountedPods. This list can be used to determine which volumes are
+	// no longer referenced and may be detached.
+	GetUnmountedVolumes() []AttachedVolume
+}
+
+// MountedVolume represents a volume that has successfully been mounted to a pod.
+type MountedVolume struct {
```

Comment:
I want the package internal components, like reconciler and populator, to operate on the package internal wrapper which can extend the type with more data. In this case the package doesn't yet need anything additional. So it's mostly for symmetry, see `AttachedVolume` and future potential.